File: test/c++/lattice/gloc.cpp
#include "nda/layout/range.hpp"
#include <algorithm>
#include <triqs/test_tools/gfs.hpp>
#include <triqs/lattice/bz_integrators.hpp>
#include <triqs/mesh/imfreq.hpp>
#include <triqs/lattice/gloc.hpp>
#include <triqs/lattice/tb_hamiltonian.hpp>

using namespace triqs;

TEST(gloc_tests, gloc_single_orbital) { // NOLINT

  // set up a tb_hamiltonian
  double t             = 1.0;
  auto displ_vec       = std::vector<std::array<long, 3>>{{1, 0, 0}, {-1, 0, 0}, {0, 1, 0}, {0, -1, 0}, {0, 0, 1}, {0, 0, -1}};
  auto overlap_mat_vec = std::vector(displ_vec.size(), nda::array<dcomplex, 2>(nda::diag(nda::vector<dcomplex>{t})));
  auto tb              = tb_hamiltonian(displ_vec, overlap_mat_vec);

  double beta  = 10;
  long n_iw    = 5;
  auto iw_mesh = mesh::imfreq{beta, gfs::Fermion, n_iw, mesh::imfreq::option::positive_frequencies_only};

  // make a zero self energy
  // TODO what is the right way to use this block2gf make function?
  //auto Sigma = gfs::gf<mesh::imfreq>{iw_mesh, {norb, norb}};

  // call the integration function
  lattice::bz_int_options options{.tolerance = 1.e-3, .k_grid = {10, 10, 10}, .k_grid_max = {20, 20, 20}};

  auto Gloc = gloc(iw_mesh, tb, 0., options);

  std::vector<dcomplex> ref = {{0, -0.419074}, {0, -0.364261}, {0, -0.316034}, {0, -0.275981}, {0, -0.243231}};
  for (auto idx_w : nda::range(n_iw)) { EXPECT_COMPLEX_NEAR(ref[idx_w], Gloc[iw_mesh[idx_w]](0, 0), 1e-3); }
}

TEST(gloc_tests, gloc_multiple_orbitals) { // NOLINT

  // set up tb_hopping object with 2 orbitals
  std::vector<std::array<long, 3>> displ_vec = {{{1, 0, 0}, {-1, 0, 0}, {0, 1, 0}, {0, -1, 0}, {0, 0, 1}, {0, 0, -1}}};
  double t                                   = 1.0;
  auto overlap_mat_vec                       = std::vector(displ_vec.size(), nda::array<dcomplex, 2>(nda::diag(nda::vector<dcomplex>{t, t})));
  auto tb                                    = tb_hamiltonian(displ_vec, overlap_mat_vec);

  double beta  = 10;
  long n_iw    = 5;
  auto iw_mesh = mesh::imfreq{beta, gfs::Fermion, n_iw, mesh::imfreq::option::positive_frequencies_only};

  // make a zero self energy
  //auto Sigma = gfs::gf<mesh::imfreq>{iw_mesh, {norb, norb}};

  // call the integration function
  lattice::bz_int_options options{.tolerance = 1.e-3, .k_grid = {10, 10, 10}, .k_grid_max = {40, 40, 40}};

  auto Gloc = gloc(iw_mesh, tb, 0., options);

  // check against reference values that this works
  std::vector<dcomplex> ref = {{0, -0.419074}, {0, -0.364261}, {0, -0.316034}, {0, -0.275981}, {0, -0.243231}};
  for (auto idx_w : nda::range(n_iw)) { EXPECT_COMPLEX_NEAR(ref[idx_w], Gloc[iw_mesh[idx_w]](0, 0), 1e-3); }
}

MAKE_MAIN;

File: test/python/experimental/gloc.py
import unittest

from triqs.experimental.lattice import TbHk, BzIntOptions, gloc
import numpy as np
from h5 import *
from triqs.mesh import MeshImFreq

class glocTests(unittest.TestCase):

  def setUp(self):

    self.displ_vec       = np.array([[1, 0, 0], [-1, 0, 0], [0, 1, 0], [0, -1, 0], [0, 0, 1], [0, 0, -1]])
    self.beta = 10.
    self.n_iw = 5
    self.iw_mesh = MeshImFreq(beta=self.beta, statistic='Fermion', n_iw=self.n_iw, opt="positive_frequencies_only")
    self.ref = [0+ -0.419074j, 0+ -0.364261j, 0+ -0.316034j, 0+ -0.275981j, 0 + -0.243231j]

  def test_gloc_single_orbital(self):

    overlap_mat_vec = np.array([np.eye(1) for _ in range(self.displ_vec.shape[0])])
    tb = TbHk(self.displ_vec, overlap_mat_vec)

    # call the integration function
    options = BzIntOptions(k_grid = [10,10,10], k_grid_max=[20,20,20], tolerance = 1.e-3)

    gloc_result = gloc(self.iw_mesh, tb, 0., options)

    for idx_w in range(self.n_iw):
      self.assertAlmostEqual(self.ref[idx_w], gloc_result[self.iw_mesh[idx_w]][0, 0], places=3)

  def test_gloc_multiple_orbitals(self):

    # two orbitals
    overlap_mat_vec = np.array([np.eye(2) for _ in range(self.displ_vec.shape[0])])
    tb = TbHk(self.displ_vec, overlap_mat_vec)

    # call the integration function
    options = BzIntOptions(k_grid = [10,10,10], k_grid_max=[40,40,40], tolerance = 1.e-3)

    gloc_result = gloc(self.iw_mesh, tb, 0., options)

    for idx_w in range(self.n_iw):
      self.assertAlmostEqual(self.ref[idx_w], gloc_result[self.iw_mesh[idx_w]][0, 0], places=3)

if __name__ == '__main__':
    unittest.main()